### PR TITLE
feat: Authenticate with API tokens

### PR DESCRIPTION
**Pull request by ian-fox**
* Users can trade their access token for a 12h JWT

Related to https://github.com/screwdriver-cd/screwdriver/issues/524

### CODE REVIEW

File: test/plugins/auth.test.js [hunk 1]
```diff
@@ -485,14 +486,34 @@ describe('auth plugin test', () => {
             })
         );
 
-        it('fails to issue a jwt given an invalid application auth token', () =>
+        it('returns user signed token given an API access token', () =>
             server.inject({
+                url: `/auth/token?access_key=${apiKey}`
```
Comment (tkyi) [inline test/plugins/auth.test.js]: Is there a possibility that the `temporaryAccessKey` and `apiKey` could collide?

Comment (ian-fox) [inline test/plugins/auth.test.js]: All of our access keys match `/[a-zA-Z0-9+/]{43}=/`, so if somebody picks an access key that matches that and it happens to be the same as one that we generate then there would be a collision.

Comment (ian-fox) [inline test/plugins/auth.test.js]: It would be very unlikely though.

Comment (FenrirUnbound) [inline test/plugins/auth.test.js]: > Is there a possibility that the temporaryAccessKey and apiKey could collide?

Yes. Once we fully support the app tokens, we should switch away from the `temporaryAccessKey`.

File: plugins/auth/index.js [hunk 2]
```diff
@@ -113,15 +113,30 @@ exports.register = (server, options, next) => {
                 accessTokenName: 'access_key',
                 allowCookieToken: false,
                 allowQueryToken: true,
-                validateFunc: (token, cb) => {
-                    if (token !== pluginOptions.temporaryAccessKey) {
-                        return cb(null, false, {});
+                validateFunc: function _validateFunc(token, cb) {
```
Comment (FenrirUnbound) [inline plugins/auth/index.js]: Why switch away from fat-arrow notation?

Comment (ian-fox) [inline plugins/auth/index.js]: Arrow notation doesn't bind the request object to `this`

The only thing I need the request object for is to get access to the userFactory though, so if there's a better way to get at that then there's no need for `function` notation.

Comment (FenrirUnbound) [inline plugins/auth/index.js]: What's different between `this.server` and `server` here? https://github.com/screwdriver-cd/screwdriver/blob/cf158a135aa11171683d9d4bc8cc58fc2e596f9b/plugins/auth/index.js#L112

File: plugins/auth/index.js [hunk 3]
```diff
@@ -114,14 +114,29 @@ exports.register = (server, options, next) => {
...
                     }
 
-                    return cb(null, true, {
-                        username: pluginOptions.temporaryAccessUser,
-                        scope: ['user']
-                    });
+                    // Token is an API token
+                    // using function syntax makes 'this' the request
+                    // TODO: Should log that we're authenticating a user with a token
+                    const factory = server.app.userFactory;
```
Comment (minz1027) [inline plugins/auth/index.js]: I remember when you try to access server.app not from the request, it is kinda weird. I tried it locally, it is actually `server._parent.app.userFactory`. Can you start the server locally and confirm?

Comment (ian-fox) [inline plugins/auth/index.js]: You're totally right. I wonder why the automated test didn't catch that.

Comment (ian-fox) [inline plugins/auth/index.js]: Accessing private properties seems kind of suspect though, would it be better to use `function` notation and get the factory from the request?

Comment (ian-fox) [inline plugins/auth/index.js]: Eslint actually complains about it too, won't let you run npm test if you try to access `server._parent`. I think moving back to `function` notation and using the request is probably the correct choice.

File: plugins/auth/index.js (hunk 2 above)
Comment (ian-fox) [inline plugins/auth/index.js]: Apparently the `server` object doesn't actually have `server.app.userFactory`, to get that you need to get `server._parent.app.userFactory`. Using a private property seems suspect, and ESLint actually won't allow it. 

Between using `server._parent` and disabling the ESLint rule for the line or using `function` notation, I feel like `function` is cleaner.

File: test/plugins/auth.test.js (hunk 1 above)
Comment (tkyi) [inline test/plugins/auth.test.js]: Ohh okay. So we can merge this PR but make sure we do away with the `temporaryAccessKey` soon after.

**Pull request merged**